CMakeLists.txt

```diff
@@ -1,171 +1,104 @@
-endif()
+# build options - documentation
+option(WITH_HTML "Build HTML documentation" ON)
+option(WITH_MAN "Build man pages" ON)
 
-# python
-if(WITH_BINDINGS)
-    if(NOT PYTHON_DESIRED)
-        find_package(PythonInterp REQUIRED)
-    elseif(${PYTHON_DESIRED} STREQUAL "2")
-        find_package(PythonInterp 2 EXACT REQUIRED)
-    elseif(${PYTHON_DESIRED} STREQUAL "3")
-        find_package(PythonInterp 3 EXACT REQUIRED)
-    elseif(EXISTS ${PYTHON_DESIRED})
-        set(PYTHON_EXECUTABLE ${PYTHON_DESIRED})
-        find_package(PythonInterp REQUIRED)
-    else()
-        message(FATAL_ERROR "Invalid PYTHON_DESIRED value: " ${PYTHON_DESIRED})
-    endif()
-    find_package(PythonLibs REQUIRED)
-endif()
+# build options - tests
+option(WITH_TESTS "Build tests" ON)
 
+# build options - debugging
+option(WITH_SANITIZERS "Build with address, leak and undefined sanitizers (DEBUG ONLY)" OFF)
 
-# valgrind path
-find_program(VALGRIND_PROGRAM NAMES valgrind PATH /usr/bin /usr/local/bin)
+# build options - bindings
+option(WITH_GO "Build Go bindings" ON)
+option(WITH_PERL5 "Build Perl 5 bindings" ON)
+option(WITH_PYTHON3 "Build Python 3 bindings" ON)
+option(WITH_RUBY "Build Ruby bindings" ON)
 
 
-# compiler options
-add_compile_options(-Wcast-align -Wno-uninitialized -Wredundant-decls -Wwrite-strings -Wformat-nonliteral -Wmissing-format-attribute -Wsign-compare -Wtype-limits -Wuninitialized -Wall -Wl,--as-needed)
-set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=gnu11 -Wmissing-prototypes -Waggregate-return -Wshadow -Werror=implicit-function-declaration")
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -Wmissing-declarations")
-
-# apple: turn rpath off
-set(CMAKE_MACOSX_RPATH 0)
-
-# package/project version
-add_definitions(-DPACKAGE_VERSION=\\"${LIBDNF_VERSION}\\")
+# includes
+include(GNUInstallDirs)
 
-# The libdnf API is under development now. This enables it for internal usage.
-add_definitions(-DLIBDNF_UNSTABLE_API)
 
-# gettext
-add_definitions(-DGETTEXT_DOMAIN=\\"libdnf\\")
-add_definitions(-DG_LOG_DOMAIN=\\"libdnf\\")
+# common dependencies
+find_package(PkgConfig REQUIRED)
 
-# gpgme: LargeFile Support is required on 32bit architectures
-add_definitions(-D_FILE_OFFSET_BITS=64)
 
-# tests
-add_definitions(-DTESTDATADIR=\\"${CMAKE_SOURCE_DIR}/data/tests\\")
+# C++ standard
+add_compile_options(-std=c++17)
 
-# librhsm
-if(ENABLE_RHSM_SUPPORT)
-    add_definitions(-DRHSM_SUPPORT)
-endif()
+# warnings
+add_compile_options(-Wall -Wextra -Werror)
+add_compile_options(-Wcast-align -Wformat-nonliteral -Wmissing-format-attribute -Wredundant-decls -Wsign-compare -Wtype-limits -Wuninitialized -Wwrite-strings)
+# not sure about the conversion warnings being errors; review later
+add_compile_options(-Wconversion)
 
-# libsolv
-if(ENABLE_SOLV_URPMREORDER)
-    add_definitions(-DLIBSOLV_FLAG_URPMREORDER=1)
+# linker options
+if(NOT CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+    add_compile_options(-Wl,--as-needed)
 endif()
```

Comment:
Is there a reason we strip this for Clang?

Reply:
Yes: clang-10: error: -Wl,--as-needed: 'linker' input unused [-Werror,-Wunused-command-line-argument]
I'm going to write a comment explaining this.